Docs: rest_api: document `processing_steps` (#1872)
dlt-hub/dlt: docs/website/docs/dlt-ecosystem/verified-sources/rest_api/basic.md

    yield from [{"name": "dlt"}, {"name": "verified-sources"}, {"name": "dlthub-education"}]
```

### Processing steps: filter and transform data

The `processing_steps` field in the resource configuration allows you to apply transformations to the data fetched from the API before it is loaded into your destination. This is useful when you need to filter out certain records, modify the data structure, or anonymize sensitive information.

> Each processing step is a dictionary specifying the type of operation (`filter` or `map`) and the function to apply.

Reviewer: ```suggestion
Each processing step is a dictionary specifying the type of operation (`filter` or `map`) and the function to apply. Steps apply in the order they are listed.
```

Reviewer: Idea: What do you think about mentioning it here already?
It's certainly clear due to the tip box given below! Not sure how much redundancy we want.

Author: I think it won't hurt :)

#### Quick example

```py
def lower_title(record):
    record["title"] = record["title"].lower()
    return record

config: RESTAPIConfig = {
    "client": {
        "base_url": "https://api.example.com",
    },
    "resources": [
        {
            "name": "posts",

>             "endpoint": "posts",

Reviewer: I'd leave this out since it's inferred from the name
```suggestion
```

            "processing_steps": [
                {"filter": lambda x: x["id"] < 10},
                {"map": lower_title},
            ],
        },
    ],
}
```

In the example above:

> - The `filter` step uses a lambda function to include only records where `id` is less than 10.
> - The `map` step applies the `lower_title` function to each remaining record.

Reviewer: ```suggestion
- First, the `filter` step uses a lambda function to include only records where `id` is less than 10.
- Thereafter, the `map` step applies the `lower_title` function to each remaining record.
```

Author: Perfect, thank you!

#### Using `filter`

The `filter` step allows you to exclude records that do not meet certain criteria. The provided function should return `True` to keep the record or `False` to exclude it:

```py
{
    "name": "posts",
    "endpoint": "posts",
    "processing_steps": [
        {"filter": lambda x: x["id"] in [10, 20, 30]},
    ],
}
```

In this example, only records with `id` equal to 10, 20, or 30 will be included.

#### Using `map`

The `map` step allows you to modify the records fetched from the API. The provided function should take a record as an argument and return the modified record. For example, to anonymize the `email` field:

```py
def anonymize_email(record):
    record["email"] = "REDACTED"
    return record

config: RESTAPIConfig = {
    "client": {
        "base_url": "https://api.example.com",
    },
    "resources": [
        {
            "name": "users",

>             "endpoint": "users",

Reviewer: I'd leave this out since it's inferred from the `name`
```suggestion
```

            "processing_steps": [
                {"map": anonymize_email},
            ],
        },
    ],
}
```

#### Combining `filter` and `map`

You can combine multiple processing steps to achieve complex transformations:

```py
{
    "name": "posts",
    "endpoint": "posts",
    "processing_steps": [
        {"filter": lambda x: x["id"] < 10},
        {"map": lower_title},
        {"filter": lambda x: "important" in x["title"]},
    ],
}
```

:::tip
#### Best practices
1. Order matters: Processing steps are applied in the order they are listed. Be mindful of the sequence, especially when combining `map` and `filter`.
2. Function definition: Define your filter and map functions separately for clarity and reuse.
3. Use `filter` to exclude records early in the process to reduce the amount of data that needs to be processed.

> 4. Combine consecutive `map` steps into a single function for better performance.

Reviewer: Maybe we can clarify that "performance" here means runtime, not memory or i/o.
```suggestion
4. Combine consecutive `map` steps into a single function for faster execution.
```

Author: yes, makes sense!
